### PR TITLE
Fix planetocentric rectangular coordinates

### PR DESCRIPTION
**Pull request by worachate001**
### Description
As mentioned in #2237 Sun & Moon are not aligned at greatest solar eclipse. I suspect that something is wrong with observer's coordinates on Earth and it's right.

There is a bug in planetocentric rectangular coordinate calculations that give wrong values of rhoCosPhiPrime and rhoSinPhiPrime. After fixing this issue, the Sun and Moon are aligned perfectly for all central solar eclipses! This should also improve the accuracy of topocentric coordinates for all objects.

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

**Test Configuration**:
* Operating system: Ubuntu 20.04

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/gui/AstroCalcDialog.cpp [hunk 1]
```diff
@@ -3107,7 +3109,10 @@ cSEparameter centralSolarEclipse(double JD) {
 	double L2 = bessel.L2;
 	const double mu = bessel.mu;
 
-	constexpr double e2 = 0.00669438;
+	static SolarSystem* ssystem = GETSTELMODULE(SolarSystem);
+	const double f = 1.0 - ssystem->getEarth()->getOneMinusOblateness(); // flattening
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:3113]: you could also make this and the next 2 static.

File: src/core/modules/Planet.cpp [hunk 2]
```diff
@@ -916,15 +916,17 @@ class SolarEclipse
 		double L1 = z * tf1 + (0.2725076 / cos(f1));
 		double L2 = z * tf2 - (0.272281 / cos(f2));
 		double mu = gast - a / M_PI_180;
-		constexpr double e2 = 0.00669438;
+		const double f = 1. - ssystem->getEarth()->getOneMinusOblateness(); // flattening
```
Comment (gzotti) [inline src/core/modules/Planet.cpp:919]: good to identify these as well. could become static, and save a multiplication with 
`e2=f*(2.-f);
`

File: src/core/modules/Planet.cpp [hunk 3]
```diff
@@ -1748,7 +1750,7 @@ Vec4d Planet::getRectangularCoordinates(const double longDeg, const double latDe
 	// For unclear reasons latDeg can be nan. Safety measure:
 	const double latRad = std::isnan(latDeg) ? 0. : latDeg*M_PI_180;
 	Q_ASSERT_X(!std::isnan(latRad), "Planet.cpp", QString("NaN result for latRad. Object %1 latitude %2").arg(englishName).arg(QString::number(latDeg, 'f', 5)).toLatin1());
-	const double u = (abs(latRad) - M_PI_2 < 1e-10 ? latRad : atan( bByA * tan(latRad)) );
+	const double u = (abs(abs(latRad) - M_PI_2) < 1e-10 ? latRad : atan( bByA * tan(latRad)) );
```
Comment (gzotti) [inline src/core/modules/Planet.cpp:1753]: Whoa! Thanks for that one!

File: src/gui/AstroCalcDialog.cpp [hunk 4]
```diff
@@ -3018,16 +3018,19 @@ pSEparameter partialSolarEclipse() {
 	double L1 = bessel.L1;
 	double L2 = bessel.L2;
 	const double mu = bessel.mu;
-	constexpr double e2 = 0.00669438;
+
+	static SolarSystem* ssystem = GETSTELMODULE(SolarSystem);
+	const double f = 1.0 - ssystem->getEarth()->getOneMinusOblateness(); // flattening
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:3023]: same as above...

File: src/gui/AstroCalcDialog.cpp [hunk 5]
```diff
@@ -3225,24 +3230,20 @@ localSEparameter localSolarEclipse(double JD,int contact,bool central) {
...
 	double lon = static_cast<double>(core->getCurrentLocation().longitude);
 	double elevation = static_cast<double>(core->getCurrentLocation().altitude);
-	lat = lat * M_PI_180;
 	lon = -(lon);
 	double L = 0.;
 
+	static SolarSystem* ssystem = GETSTELMODULE(SolarSystem);
+	//PlanetP earth = ssystem->getEarth();
+	Vec4d geocentricCoords = ssystem->getEarth()->getRectangularCoordinates(lon,lat,elevation);
+	const double rc = geocentricCoords[0]/(EARTH_RADIUS/AU); // rhoCosPhiPrime
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:3239]: not sure where EARTH_RADIUS is set here. preferrably `ssystem->getEarth->getEquatorialRadius()` somewhere?

**Comment by worachate001** [comment]
@gzotti Thanks.

File: src/core/modules/Planet.cpp (hunk 3 above)
Comment (worachate001) [reply src/core/modules/Planet.cpp:1753]: Yeah, take some times to realized that.

File: src/gui/AstroCalcDialog.cpp (hunk 5 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp]: Replaced with getEquatorialRadius().

File: src/gui/AstroCalcDialog.cpp [hunk 6]
```diff
@@ -3225,24 +3230,19 @@ localSEparameter localSolarEclipse(double JD,int contact,bool central) {
 	double lat = static_cast<double>(core->getCurrentLocation().latitude);
 	double lon = static_cast<double>(core->getCurrentLocation().longitude);
 	double elevation = static_cast<double>(core->getCurrentLocation().altitude);
-	lat = lat * M_PI_180;
 	lon = -(lon);
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:3233]: Not sure what this does. Is `lon` use further down? (It looks like in the previous solution it was required to switch from our eastern-positive location to an western-positive implementation?)

Else, very good that you can identify and clean up all these bits and pieces, thank you!

File: src/gui/AstroCalcDialog.cpp [hunk 7]
```diff
@@ -3225,24 +3230,19 @@ localSEparameter localSolarEclipse(double JD,int contact,bool central) {
...
 	double L = 0.;
 
+	static SolarSystem* ssystem = GETSTELMODULE(SolarSystem);
+	Vec4d geocentricCoords = ssystem->getEarth()->getRectangularCoordinates(lon,lat,elevation);
+	static const double earthRadius = ssystem->getEarth()->getEquatorialRadius();
+	static const double rc = geocentricCoords[0]/earthRadius; // rhoCosPhiPrime
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.cpp:3239]: Uh, `rc `and `rs `must not be static! When you change location, the `geocentricCoords` should also change! (static here means "set at first run and keep results until program exits")

File: src/gui/AstroCalcDialog.cpp (hunk 6 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp]: Uhmm. I think we still need to do that.

File: src/gui/AstroCalcDialog.cpp (hunk 7 above)
Comment (worachate001) [reply src/gui/AstroCalcDialog.cpp]: Oops. fixed. It should be okay now.

**Comment by alex-w** [comment]
Part of code in Planes.cpp and AstroCalcDialog.cpp files are identical - why not reuse it?

**Comment by worachate001** [comment]
> Part of code in Planes.cpp and AstroCalcDialog.cpp files are identical - why not reuse it?

Yes. Thinking about that, but I need a rest soon, it's almost midnight here.

**Comment by worachate001** [comment]
Update: the revision is going well until I found a new bug in AstroCalc/Eclipses. Trying to fix it before pushing this commit.

**Comment by alex-w** [comment]
Why not use official term instead of "Sun/Moon diameter ratio"?

**Comment by worachate001** [comment]
> Why not use official term instead of "Sun/Moon diameter ratio"?

magnitude-ratio-tse

This is an example of total eclipse when you're standing near the limit of umbra.

We already have 'Eclipse magnitude' below 'Eclipse obscuration' - eclipse magnitude here indicates how close you are to the northern or southern limit of umbra. It will be close to 1.0 if you are just inside the umbra.

Moon/Sun diameter ratio is different value, if we use the same term, it will be contradict to Eclipse magnitude above it. I think users may be confused of having two magnitudes at the same time.

**Comment by worachate001** [comment]
Example of magnitude at greatest eclipse from two sources that show different value for the same event (look at 1955 Jun 20)

NASA

and 
https://astro.ukho.gov.uk/eclipse/0311955/Bangkok_Thailand_1955Jun20.png

**Comment by gzotti** [comment]
Probably we should re-arrange the lines, though:

* Eclipse magnitude
* Eclipse obscuration
* Moon/Sun diameter ratio
* Central eclipse duration
* ...

By this, we would have the "coverage" items first, then the geographic data.

File: src/core/modules/Planet.cpp [hunk 8]
```diff
@@ -1427,6 +1407,25 @@ QString Planet::getInfoStringExtra(const StelCore *core, const InfoStringGroup&
...
+
+						oss << QString("%1 %2 %3 %4°<br/>").arg(
+								 q_("Shadow center point is"),
+								 QString::number(distance, 'f', 1),
+								 q_("km towards azimuth"),
+								 QString::number(azimuth, 'f', 1));
+						oss << QString("%1: %2 %3<br/>").arg(
+								 q_("Width of umbra/antumbra"),
+								 QString::number(pathWidth, 'f', 1),
+								 q_("km"));
```
Comment (alex-w) [inline src/core/modules/Planet.cpp:1428]: I think variable `km` with context data was defined before

**Review by alex-w — APPROVED** [review]
Thanks for improving!

File: src/core/modules/Planet.cpp (hunk 8 above)
Comment (gzotti) [reply src/core/modules/Planet.cpp:1428]: yes, please use qs_("km", distance")

Comment (alex-w) [reply src/core/modules/Planet.cpp:1428]: A typo: `qc_(“term”, “context”);`

File: src/core/modules/Planet.cpp [hunk 9]
```diff
@@ -1275,34 +1379,53 @@ QString Planet::getInfoStringExtra(const StelCore *core, const InfoStringGroup&
...
-						oss << QString("%1: %2 ").arg(
-								 q_("Magnitude of central eclipse"),
-								 QString::number(pos[2], 'f', 3));
-						if (pos[2] < 1.0)
-							oss << QString(qc_("(annular)","type of solar eclipse"));
-						else
-							oss << QString(qc_("(total)","type of solar eclipse"));
-						oss << "<br/>";
+						oss << QString("%1: %2 %3<br/>").arg(
+								 q_("Width of umbra/antumbra"),
```
Comment (gzotti) [inline src/core/modules/Planet.cpp:1426]: Can you separate and give two separate words: either "umbra" (for total eclipses or total phase of hybrid) or "antumbra" (annular/annular phase of hybrid)?

File: src/core/modules/Planet.cpp (hunk 8 above)
Comment (gzotti) [reply src/core/modules/Planet.cpp:1428]: Argh, sorry. qc_...

**Pull request merged**